lib/ABI/FunctionType/Layout.cpp

```diff
@@ -0,0 +1,963 @@
+  DistributedArguments Result;
+
+  std::size_t UsedVectorRegisterCounter = 0;
+  std::size_t UsedGeneralPurposeRegisterCounter = SkippedCount;
+  for (const model::Argument &Argument : Arguments) {
+    bool CanSplit = ABI.ArgumentsCanBeSplitBetweenRegistersAndStack();
+
+    std::size_t RegisterLimit = 0;
+    std::size_t *RegisterCounter = nullptr;
+    std::span<const model::Register::Values> RegisterList;
+    if (Argument.Type().isFloat()) {
+      RegisterList = ABI.VectorArgumentRegisters();
+      RegisterCounter = &UsedVectorRegisterCounter;
+
+      // The conventional non-position based approach is not applicable for
+      // vector registers since it's rare for multiple registers to be used
+      // to pass a single argument.
+      //
+      // For now, provide at most a single vector register for such a value,
+      // if there's a free one.
+      //
+      // TODO: find reproducers and handle the cases where multiple vector
+      //       registers are used together.
+      RegisterLimit = 1;
+
+      // Explicitly disallow splitting vector arguments across the registers
+      // and the stack.
+      CanSplit = false;
+    } else {
+      RegisterList = ABI.GeneralPurposeArgumentRegisters();
+      RegisterCounter = &UsedGeneralPurposeRegisterCounter;
+      RegisterLimit = Argument.Type().isScalar() ?
+                        ABI.MaximumGPRsPerScalarArgument() :
+                        ABI.MaximumGPRsPerAggregateArgument();
+    }
+
+    auto [Arguments, NextRegisterIndex] = distributeImpl(Argument.Type(),
+                                                         RegisterList,
+                                                         *RegisterCounter,
+                                                         RegisterLimit,
+                                                         CanSplit);
+
+    // Verify that the next register makes sense.
+    auto VerifyNextRegisterIndex = [&](std::size_t Current, std::size_t Next) {
+      if (Current == Next)
+        return true; // No registers were used for this argument.
+
+      if (Next >= Current && Next <= Current + RegisterLimit)
+        return true; // It's within the expected boundaries.
+
+      if (Next == RegisterList.size()) {
+        // All the register are marked as used. Only allow this on ABIs that
+        // don't allow register arguments to come after stack ones.
+        return ABI.NoRegisterArgumentsCanComeAfterStackOnes();
+      }
+
+      return false;
+    };
+    revng_assert(VerifyNextRegisterIndex(*RegisterCounter, NextRegisterIndex));
+    *RegisterCounter = NextRegisterIndex;
+
+    // All good - add the argument.
+    auto Movable = llvm::make_range(std::make_move_iterator(Arguments.begin()),
+                                    std::make_move_iterator(Arguments.end()));
+    for (DistributedArgument &&Argument : Movable)
+      Result.emplace_back(std::move(Argument));
```

Comment:
Frankly this makes the code obscure for a totally dubious return.
At very least, outline the pattern `make_range(make_move_iterator(...), make_move_iterator(...))` in its own function, but this is a clear example of early optimization and a bad trade off between complexity and return.

Reply:
Moved it out and gave it the name from the STL so eventually we can collapse it onto the standard library once we update our implementation (see `https://github.com/revng/revng/pull/232/commits/3409eb763a7893e94c58a0cfc724b4a033c94d90`)